File: experiment-5/runs/language=typescript_model=claude-opus-4-7_tooling=none/rep3/src/queries/stats.ts
import { Match, Competition } from '../types.js';

export interface OverallStats {
  matches: number;
  totalGoals: number;
  averageGoals: number;
  homeWins: number;
  awayWins: number;
  draws: number;
  homeWinRate: number;
  awayWinRate: number;
  drawRate: number;
}

export function overallStats(
  matches: Match[],
  opts: { competition?: Competition; season?: number } = {},
): OverallStats {
  const filtered = matches.filter((m) => {
    if (opts.competition && m.competition !== opts.competition) return false;
    if (opts.season != null && m.season !== opts.season) return false;
    return true;
  });

  let totalGoals = 0;
  let homeWins = 0;
  let awayWins = 0;
  let draws = 0;
  for (const m of filtered) {
    totalGoals += m.homeGoals + m.awayGoals;
    if (m.homeGoals > m.awayGoals) homeWins++;
    else if (m.homeGoals < m.awayGoals) awayWins++;
    else draws++;
  }
  const n = filtered.length;
  return {
    matches: n,
    totalGoals,
    averageGoals: n > 0 ? totalGoals / n : 0,
    homeWins,
    awayWins,
    draws,
    homeWinRate: n > 0 ? homeWins / n : 0,
    awayWinRate: n > 0 ? awayWins / n : 0,
    drawRate: n > 0 ? draws / n : 0,
  };
}

export interface SeasonsAvailable {
  season: number;
  matches: number;
}

export function seasonsAvailable(
  matches: Match[],
  competition?: Competition,
): SeasonsAvailable[] {
  const counts = new Map<number, number>();
  for (const m of matches) {
    if (competition && m.competition !== competition) continue;
    if (m.season == null) continue;
    counts.set(m.season, (counts.get(m.season) ?? 0) + 1);
  }
  return Array.from(counts.entries())
    .map(([season, matches]) => ({ season, matches }))
    .sort((a, b) => a.season - b.season);
}

File: experiment-14/runs/language=elixir_model=opus-4.8-fast_prompt=neutral/rep1/lib/br_soccer/stats.ex
defmodule BrSoccer.Stats do
  @moduledoc "Aggregate statistics over filtered match sets."

  alias BrSoccer.{Match, Matches}

  @doc """
  Aggregate stats for a filtered match set (same options as
  `BrSoccer.Matches.search/1`).

  Returns counts, average goals per match, home/away/draw win rates and goal
  totals.
  """
  def summary(opts \\ []) do
    matches = Matches.search(Keyword.put(opts, :scored_only, true))
    count = length(matches)

    {home, away, draw, goals} =
      Enum.reduce(matches, {0, 0, 0, 0}, fn m, {h, a, d, g} ->
        total = m.home_goal + m.away_goal

        case Match.result(m) do
          :home_win -> {h + 1, a, d, g + total}
          :away_win -> {h, a + 1, d, g + total}
          :draw -> {h, a, d + 1, g + total}
          _ -> {h, a, d, g}
        end
      end)

    %{
      matches: count,
      total_goals: goals,
      avg_goals: ratio(goals, count),
      home_wins: home,
      away_wins: away,
      draws: draw,
      home_win_rate: pct(home, count),
      away_win_rate: pct(away, count),
      draw_rate: pct(draw, count)
    }
  end

  @doc "Compare two seasons of a competition side by side."
  def compare_seasons(competition, season_a, season_b) do
    %{
      competition: competition,
      season_a: season_a,
      season_b: season_b,
      a: summary(competition: competition, season: season_a),
      b: summary(competition: competition, season: season_b)
    }
  end

  defp ratio(_n, 0), do: 0.0
  defp ratio(n, total), do: Float.round(n / total, 2)

  defp pct(_n, 0), do: 0.0
  defp pct(n, total), do: Float.round(n * 100 / total, 1)
end
